File: rust/src/bin/get-block.rs
//! Reads a block containing v1 transactions over JSON-RPC.
//!
//! When `maxSupportedTransactionVersion` is below the highest version present,
//! the whole request fails rather than degrading, so a caller pinned at 0 loses
//! entire blocks as soon as v1 traffic appears.

use anyhow::{Context, Result};
use clap::Parser;
use solana_commitment_config::CommitmentConfig;
use solana_keypair::Keypair;
use solana_rpc_client::rpc_client::RpcClient;
use solana_rpc_client_api::config::RpcBlockConfig;
use solana_signer::Signer;
use solana_transaction::versioned::TransactionVersion;
use solana_transaction_status_client_types::{
    EncodedTransaction, TransactionDetails, UiConfirmedBlock, UiMessage, UiTransactionEncoding,
};
use txv1::send::{build_v1_transfer, funded_keypair, EXAMPLE_CONFIG};
use txv1::version::VersionTally;
use txv1::{MessageVersion, DEFAULT_RPC_URL};

#[derive(Parser)]
#[command(about = "Fetch a block and tally its transaction versions")]
struct Args {
    #[arg(long, env = "TXV1_RPC_URL", default_value = DEFAULT_RPC_URL)]
    rpc_url: String,

    /// Slot to read. Defaults to the slot of a freshly sent v1 transaction.
    #[arg(long)]
    slot: Option<u64>,
}

fn main() -> Result<()> {
    let args = Args::parse();
    let client = RpcClient::new_with_commitment(args.rpc_url, CommitmentConfig::confirmed());

    let slot = match args.slot {
        Some(slot) => slot,
        None => send_v1_and_get_slot(&client)?,
    };
    println!("reading slot {slot}\n");

    println!("== maxSupportedTransactionVersion omitted ==");
    match client.get_block_with_config(slot, block_config(None)) {
        Ok(_) => println!("  succeeded: this block holds no versioned transactions"),
        Err(error) => println!("  rejected: {error}"),
    }

    println!("\n== maxSupportedTransactionVersion: 0 ==");
    match client.get_block_with_config(slot, block_config(Some(0))) {
        Ok(_) => println!("  succeeded: this block holds no v1 transactions"),
        Err(error) => println!("  rejected: {error}"),
    }

    println!("\n== maxSupportedTransactionVersion: 1 ==");
    let block = client
        .get_block_with_config(slot, block_config(Some(1)))
        .context("getBlock with maxSupportedTransactionVersion: 1 failed")?;
    report(&block);

    Ok(())
}

fn block_config(max_supported_transaction_version: Option<u8>) -> RpcBlockConfig {
    RpcBlockConfig {
        commitment: Some(CommitmentConfig::confirmed()),
        encoding: Some(UiTransactionEncoding::Json),
        max_supported_transaction_version,
        rewards: Some(false),
        transaction_details: Some(TransactionDetails::Full),
    }
}

fn send_v1_and_get_slot(client: &RpcClient) -> Result<u64> {
    let payer = funded_keypair(client, 1_000_000_000)?;
    let recipient = Keypair::new().pubkey();
    let blockhash = client.get_latest_blockhash()?;
    let transaction = build_v1_transfer(&payer, &recipient, 10_000_000, blockhash, EXAMPLE_CONFIG)?;
    let signature = client.send_and_confirm_transaction(&transaction)?;
    let statuses = client.get_signature_statuses(&[signature])?;
    let status = statuses
        .value
        .into_iter()
        .flatten()
        .next()
        .context("no status for the transaction just sent")?;
    Ok(status.slot)
}

fn report(block: &UiConfirmedBlock) {
    let Some(transactions) = &block.transactions else {
        println!("  block carried no transaction details");
        return;
    };

    let mut tally = VersionTally::default();
    let mut v1_lines = Vec::new();

    for transaction in transactions {
        // Unlike gRPC, JSON-RPC reports the version — but only because the
        // request opted in.
        let version = match transaction.version {
            Some(TransactionVersion::Number(1)) => MessageVersion::V1,
            Some(TransactionVersion::Number(0)) => MessageVersion::V0,
            Some(TransactionVersion::Number(other)) => {
                // Folding an unrecognised version into an existing bucket is
                // how v1 first shows up as v0 in an unprepared pipeline.
                println!("  warning: unrecognised transaction version {other}, not tallied");
                continue;
            }
            _ => MessageVersion::Legacy,
        };
        tally.record(version);

        if version != MessageVersion::V1 {
            continue;
        }
        let EncodedTransaction::Json(ui) = &transaction.transaction else {
            continue;
        };
        let config = match &ui.message {
            UiMessage::Raw(raw) => raw.transaction_config.as_ref(),
            UiMessage::Parsed(parsed) => parsed.transaction_config.as_ref(),
        };
        let signature = ui.signatures.first().map_or("?", String::as_str);
        v1_lines.push(match config {
            Some(config) => format!(
                "  {signature}\n    priority_fee={:?} compute_unit_limit={:?} \
                 loaded_accounts_data_size_limit={:?} heap_size={:?}",
                config.priority_fee,
                config.compute_unit_limit,
                config.loaded_accounts_data_size_limit,
                config.heap_size
            ),
            None => format!("  {signature}\n    no transactionConfig on a v1 transaction"),
        });
    }

    println!("  {} of {} transactions", tally, tally.total());
    for line in v1_lines {
        println!("{line}");
    }
}

File: ts/src/get-block.ts
/**
 * Reads a block containing v1 transactions over JSON-RPC.
 *
 * When `maxSupportedTransactionVersion` is below the highest version in the
 * block, the entire request fails — a caller pinned at 0 loses whole blocks.
 *
 * Run with `just ts-get-block`.
 */

import { BLOCK_CONFIG, formatTransactionConfig } from './lib/rpc';
import { createClients, sendV1TransferAndGetSlot } from './lib/send';
import { decodeBase64Transaction } from './lib/v1';

const clients = createClients();
const slotArgument = process.argv[2];
const slot = slotArgument === undefined ? (await sendV1TransferAndGetSlot(clients)).slot : BigInt(slotArgument);
console.log(`reading slot ${slot}\n`);

const describeRejection = (error: unknown) => `  rejected: ${error instanceof Error ? error.message : String(error)}`;

console.log('== maxSupportedTransactionVersion omitted ==');
try {
    await clients.rpc.getBlock(slot, BLOCK_CONFIG).send();
    console.log('  succeeded: this block holds no versioned transactions');
} catch (error) {
    console.log(describeRejection(error));
}

console.log('\n== maxSupportedTransactionVersion 0 ==');
try {
    await clients.rpc.getBlock(slot, { ...BLOCK_CONFIG, maxSupportedTransactionVersion: 0 }).send();
    console.log('  succeeded: this block holds no v1 transactions');
} catch (error) {
    console.log(describeRejection(error));
}

console.log('\n== maxSupportedTransactionVersion 1 ==');
const block = await clients.rpc.getBlock(slot, { ...BLOCK_CONFIG, maxSupportedTransactionVersion: 1 }).send();
if (block === null || !('transactions' in block)) {
    throw new Error(`slot ${slot} has no transaction details`);
}

const tally = { legacy: 0, v0: 0, v1: 0 };
for (const transaction of block.transactions) {
    // Unlike gRPC, JSON-RPC reports the version — but only because the request
    // opted in.
    if (transaction.version === 1) {
        tally.v1 += 1;
    } else if (transaction.version === 0) {
        tally.v0 += 1;
    } else {
        tally.legacy += 1;
    }
}
console.log(`  legacy=${tally.legacy} v0=${tally.v0} v1=${tally.v1} of ${block.transactions.length} transactions`);

for (const transaction of block.transactions) {
    if (transaction.version !== 1) {
        continue;
    }
    const decoded = decodeBase64Transaction(transaction.transaction[0]);
    console.log(`  v1 ${decoded.signature}`);
    console.log(formatTransactionConfig(decoded.config, '    '));
}
